### PR TITLE
Issuer and CredentialTypes extrinsics

### PR DESCRIPTION
**Pull request by lana-shanghai**
Solves: #47

### CODE REVIEW

File: pallets/did/src/lib.rs [hunk 1]
```diff
@@ -114,88 +116,146 @@ pub mod pallet {
...
 			Ok(())
 		}
 
+		#[pallet::call_index(1)]
 		#[pallet::weight(1000000)]
 		pub fn add_issuer(origin: OriginFor<T>, issuer: DidIdentifierOf<T>) -> DispatchResult {
 			// Origin ONLY GovernanceOrigin
 			T::GovernanceOrigin::ensure_origin(origin)?;
 			// Add issuer to database with status Active
+			Issuers::<T>::insert(issuer.clone(), IssuerInfo { status: IssuerStatus::Active });
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:167]: Need to check if the Issuer already exist. If already exists then throw an error. Becase of this I realised we need an `activate_issuer` to give the possibility to set back to `Active` revoked Issuers

File: pallets/did/src/lib.rs [hunk 2]
```diff
@@ -114,88 +116,146 @@ pub mod pallet {
...
+			}
+			CredentialsTypes::<T>::put(credentials_types);
+			Self::deposit_event(Event::CredentialTypeRemoved);
 			Ok(())
 		}
 	}
+
+	impl<T: Config> Pallet<T> {
+		/// Delete issuer from Storage.
+		pub fn do_delete_issuer(issuer: DidIdentifierOf<T>) -> DispatchResult {
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:248]: Check the Issuer exists and if its status is `Revoked`. We do not allow deleting `Active` Issuers.
Throw two different errors for each case

File: pallets/did/src/lib.rs [hunk 3]
```diff
@@ -114,88 +116,146 @@ pub mod pallet {
...
+			credential: CredentialOf<T>,
+		) -> DispatchResult {
+			// Origin ONLY GovernanceOrigin
 			T::GovernanceOrigin::ensure_origin(origin)?;
+			let mut credentials_types = CredentialsTypes::<T>::get();
+			ensure!(
+				!Self::credential_exists(&credentials_types, &credential),
+				Error::<T>::CredentialAlreadyAdded
+			);
+			let pos = credentials_types.len();
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:211]: `pos` should be the location returned by the `binary_search` to keep the vector ordered, not the last position (`.len()`)

File: pallets/did/src/lib.rs [hunk 4]
```diff
@@ -114,88 +116,146 @@ pub mod pallet {
...
 			Ok(())
 		}
 
+		#[pallet::call_index(4)]
 		#[pallet::weight(1000000)]
-		pub fn add_credential_type(origin: OriginFor<T>, credential: CredentialOf<T>) -> DispatchResult {
- 			// Origin ONLY GovernanceOrigin
+		pub fn add_credential_type(
+			origin: OriginFor<T>,
+			credential: CredentialOf<T>,
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:202]: Can you check if extrinsic fails when Credential type string length is exceeded? I am not sure. Maybe better to let it as unbounded Vec and check range when received

File: pallets/did/src/lib.rs [hunk 5]
```diff
@@ -114,88 +116,146 @@ pub mod pallet {
...
+			ensure!(
+				!Self::credential_exists(&credentials_types, &credential),
+				Error::<T>::CredentialAlreadyAdded
+			);
+			let pos = credentials_types.len();
+			credentials_types
+				.try_insert(pos, credential)
+				.map_err(|_| Error::<T>::MaxCredentials)?;
+			CredentialsTypes::<T>::put(credentials_types);
+			Self::deposit_event(Event::CredentialTypeAdded);
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:216]: Please add the credential type added, it can be useful for the Indexing Service

File: pallets/did/src/lib.rs [hunk 6]
```diff
@@ -114,88 +116,146 @@ pub mod pallet {
...
+			credential: CredentialOf<T>,
+		) -> DispatchResult {
 			// Origin ONLY GovernanceOrigin
 			T::GovernanceOrigin::ensure_origin(origin)?;
+			let mut credentials_types = CredentialsTypes::<T>::get();
+			ensure!(
+				Self::credential_exists(&credentials_types, &credential),
+				Error::<T>::CredentialDoesNotExist
+			);
+			match credentials_types.binary_search(&credential) {
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:233]: duplicated `binary_search`

File: pallets/did/src/lib.rs [hunk 7]
```diff
@@ -114,88 +116,146 @@ pub mod pallet {
...
+				Ok(pos) => {
+					credentials_types.remove(pos);
+					()
+				},
+				_ => (),
+			}
+			CredentialsTypes::<T>::put(credentials_types);
+			Self::deposit_event(Event::CredentialTypeRemoved);
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:241]: Add Credential type removed to the event

File: pallets/did/src/lib.rs [hunk 8]
```diff
@@ -114,88 +116,146 @@ pub mod pallet {
...
+		pub fn do_delete_issuer(issuer: DidIdentifierOf<T>) -> DispatchResult {
+			Issuers::<T>::remove(issuer);
+			Ok(())
+		}
+
+		/// Check that a credential already exists.
+		fn credential_exists(
+			credentials: &BoundedVec<CredentialOf<T>, T::MaxCredentialsTypes>,
+			credential: &CredentialOf<T>,
+		) -> bool {
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:257]: Following previous comment about location return by `binary_search`, probably modify the method to return a `Result` with that location index.

File: pallets/did/src/types.rs [hunk 9]
```diff
@@ -34,17 +34,18 @@ pub struct Document<T: Config> {
 
 #[derive(Clone, Decode, Encode, PartialEq, TypeInfo, MaxEncodedLen)]
 pub enum IssuerStatus {
+	New,
```
Comment (NachoPal) [inline pallets/did/src/types.rs:37]: When is going to be used `New`? Why we need it?

**Review by NachoPal — CHANGES_REQUESTED** [review]

File: pallets/did/src/lib.rs (hunk 1 above)
Comment (lana-shanghai) [reply pallets/did/src/lib.rs:167]: Agree, it is best to separate the creation of a completely new user and a re-activation

File: pallets/did/src/lib.rs [hunk 10]
```diff
@@ -114,87 +116,168 @@ pub mod pallet {
...
+			CredentialsTypes::<T>::put(credentials_types);
+			Self::deposit_event(Event::CredentialTypeRemoved { credential });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Delete issuer from Storage.
+		pub fn do_delete_issuer(issuer: DidIdentifierOf<T>) -> DispatchResult {
+			Issuers::<T>::remove(issuer);
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:280]: This method is dangerous as there are not any checks inside. Add the logic for deleting the issuer inside it.

File: pallets/did/src/lib.rs [hunk 11]
```diff
@@ -114,87 +116,168 @@ pub mod pallet {
...
+			let pos = credentials_types.binary_search(&credential);
+			ensure!(!pos.is_ok(), Error::<T>::CredentialAlreadyAdded);
+			match pos {
+				Err(pos) => {
+					credentials_types
+						.try_insert(pos, credential.clone())
+						.map_err(|_| Error::<T>::MaxCredentials)?;
+					()
+				},
+				Ok(_pos) => (),
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:246]:
```suggestion
						let pos = credentials_types.binary_search(&credentials).err().ok_or(Error::<T>::CredentialAlreadyAdded)?;
						credentials_types
							.try_insert(pos, credential.clone())
							.map_err(|_| Error::<T>::MaxCredentials)?;
					}
```

File: pallets/did/src/lib.rs [hunk 12]
```diff
@@ -114,87 +116,168 @@ pub mod pallet {
...
+			let mut credentials_types = CredentialsTypes::<T>::get();
+			let pos = credentials_types.binary_search(&credential);
+			ensure!(pos.is_ok(), Error::<T>::CredentialDoesNotExist);
+			match pos {
+				Ok(pos) => {
+					credentials_types.remove(pos);
+					()
+				},
+				Err(_pos) => (),
+			}
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:270]:
```suggestion
let pos = credentials_types.binary_search(&credentials).ok().ok_or(Error::<T>::CredentialDoesNotExist)?;
credentials_types.remove(pos);
```

File: pallets/did/src/types.rs (hunk 9 above)
Comment (NachoPal) [reply pallets/did/src/types.rs:37]: > Regarding the New status, I was thinking, can there be a situation when we mutate a storage and it falls back on a Default value Active which gives more rights than a blank New status. I can remove it as well.

Yeah, better remove it for now as I do not see a clear usage of it. I think for now `Active` and `Revoked` is enough.

**Review by NachoPal — CHANGES_REQUESTED** [review]

File: pallets/did/src/lib.rs [hunk 13]
```diff
@@ -114,87 +116,177 @@ pub mod pallet {
...
+				ensure!(
+					*info == IssuerInfo { status: IssuerStatus::Revoked },
+					Error::<T>::IssuerNotRevoked
+				);
+				// Remove issuer from storage
+				*maybe_info = None;
+				Self::deposit_event(Event::IssuerDeleted { issuer });
+				Ok(())
+			})?;
 			Ok(())
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:240]:
```suggestion
			})
```

File: pallets/did/src/lib.rs [hunk 14]
```diff
@@ -114,87 +116,177 @@ pub mod pallet {
...
+				*info = IssuerInfo { status: IssuerStatus::Active };
+				Self::deposit_event(Event::IssuerStatusReactived { issuer });
+				Ok(())
+			})?;
+			Ok(())
+		}
+
+		#[pallet::call_index(4)]
 		#[pallet::weight(1000000)]
 		pub fn delete_issuer(origin: OriginFor<T>, issuer: DidIdentifierOf<T>) -> DispatchResult {
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:225]:
```suggestion
		pub fn remove_issuer(origin: OriginFor<T>, issuer: DidIdentifierOf<T>) -> DispatchResult {
```
Just to be consistent with the rest of the extrinsics naming

File: pallets/did/src/lib.rs [hunk 15]
```diff
@@ -114,87 +116,177 @@ pub mod pallet {
...
+				.ok_or(Error::<T>::CredentialDoesNotExist)?;
+			credentials_types.remove(pos);
+			CredentialsTypes::<T>::put(credentials_types);
+			Self::deposit_event(Event::CredentialTypeRemoved { credential });
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn do_delete_issuer(origin: OriginFor<T>, issuer: DidIdentifierOf<T>) -> DispatchResult {
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:286]:
```suggestion
		pub fn do_remove_issuer(origin: OriginFor<T>, issuer: DidIdentifierOf<T>) -> DispatchResult {
```
Just to be consistent with the rest of extrinsics naming

File: pallets/did/src/lib.rs [hunk 16]
```diff
@@ -114,87 +116,177 @@ pub mod pallet {
...
+				let mut info = maybe_info.as_mut().ok_or(Error::<T>::IssuerDoesNotExist)?;
+				ensure!(
+					*info == IssuerInfo { status: IssuerStatus::Revoked },
+					Error::<T>::IssuerNotRevoked
+				);
+				// Remove issuer from storage
+				*maybe_info = None;
+				Self::deposit_event(Event::IssuerDeleted { issuer });
+				Ok(())
+			})?;
```
Comment (NachoPal) [inline pallets/did/src/lib.rs:239]: All this should be moved to `do_remove_issuer()`. `remove_issuer()` will call then `do_remove_issuer()`

**Review by NachoPal — CHANGES_REQUESTED** [review]

**Review by NachoPal — APPROVED** [review]
#51 should be handled after #49 is merged. The reason of the dependancy is to make use of `ensure_controller()`

**Pull request merged**